testing/test_config.py

```diff
@@ -264,6 +264,64 @@ def pytest_addoption(parser):
         assert len(l) == 2
         assert l == ["456", "123"]
 
+
+class TestConfigFromdictargs:
+    def test_basic_behavior(self):
+        from _pytest.config import Config
+        option_dict = {
+            'verbose': 444,
+            'foo': 'bar',
+            'capture': 'no',
+        }
+        args = ['a', 'b']
+
+        config = Config.fromdictargs(option_dict, args)
+        with pytest.raises(AssertionError):
+            config.parse(['should refuse to parse again'])
+        assert config.option.verbose == 444
+        assert config.option.foo == 'bar'
+        assert config.option.capture == 'no'
+        assert config.args == args
+
+    def test_origargs(self):
+        """Show that fromdictargs can handle args in their "orig" format"""
+        from _pytest.config import Config
+        option_dict = {}
+        args = ['-vvvv', '-s', 'a', 'b']
+
+        config = Config.fromdictargs(option_dict, args)
+        assert config.args == ['a', 'b']
+        assert config._origargs == args
+        assert config.option.verbose == 4
+        assert config.option.capture == 'no'
+
+    def test_inifilename(self, tmpdir):
+        tmpdir.join("foo/bar.ini").ensure().write(py.code.Source("""
+            [pytest]
+            name = value
+        """))
+
+        from _pytest.config import Config
+        inifile = '../../foo/bar.ini'
+        option_dict = {
+            'inifilename': inifile,
+            'capture': 'no',
+        }
+
+        cwd = tmpdir.join('a/b')
+        with cwd.ensure(dir=True).as_cwd():
+            #import pudb; pudb.set_trace()
+            config = Config.fromdictargs(option_dict, ())
+
+        assert config.args == [str(cwd)]
+        assert config.option.inifilename == inifile
+        assert config.option.capture == 'no'
+
+        # this indicates this is the file used for getting configuration values
+        assert config.inifile == inifile
+        assert config.inicfg.get('name') == 'value'
```

Comment:
This looks good, but perhaps another good test would be worth adding:

1. Create `tmpdir/pytest.ini` with:

  ```ini
  [pytest]
  opt1 = default
  opt2 = value2
  ```

2. Create `tmpdir/a/b/bar.ini` with:

  ```ini
  [pytest]
  opt1 = bar
  ```

3. `Config.fromdictargs` with `inifilename` set to `tmpdir/a/b/bar.ini`, using `tmpdir` as cwd:
4. Ensure `config.option.opt1 == "bar"` and that `config.option` does not have `opt2` option.

This test guarantees that `tmpdir/pytest.ini` is not being read because of the `inifilename` option in the dict given to `fromdictargs`.